src/SessionManager/ProgramSettings.cc

```diff
@@ -34,6 +38,141 @@ void applyOptionalArgument(T& val, const string& argument_name, const cxxopts::P
+            "settings");
+        spdlog::info("If you want to ignore system settings, use --no_user_config as command-line argument.");
+        settings.merge_patch(user_settings); // user on top of system
+    }
+
+    if (system_settings_json_exists || user_settings_json_exists) {
+        settings.merge_patch(command_line_settings); // force command-line on top of user and sytem
+        SetSettingsFromJSON(settings);
+    }
+}
+
+json ProgramSettings::JSONConfigSettings(const std::string& json_file_path) {
+    std::ifstream ifs(json_file_path, std::ifstream::in);
+    json j;
+    try {
+        j = json::parse(ifs);
+    } catch (json::exception err) {
+        spdlog::warn("Config file {} has problems, please check", json_file_path);
+        spdlog::warn(err.what());
+    }
+    return j;
+}
+
+void ProgramSettings::SetSettingsFromJSON(const json& j) {
+    ValidateJSON(j); // check for invalid keys, store them in std::set for easy counting
+    // logic here is
+    // value = check invalid, if invalid then default value, otherwise check json and assign its value, otherwise default
+    if (!invalid_keys.count("verbosity")) {
+        verbosity = j.value("verbosity", verbosity);
+    }
+    if (!invalid_keys.count("no_log")) {
+        no_log = j.value("no_log", no_log);
+    }
+    if (!invalid_keys.count("log_performance")) {
+        log_performance = j.value("log_performance", log_performance);
+    }
+    if (!invalid_keys.count("log_protocol_messages")) {
+        log_protocol_messages = j.value("log_protocol_messages", log_protocol_messages);
+    }
+    if (!invalid_keys.count("no_http")) {
+        no_http = j.value("no_http", no_http);
+    }
+    if (!invalid_keys.count("no_browser")) {
+        no_browser = j.value("no_browser", no_browser);
+    }
+    if (!invalid_keys.count("host")) {
+        host = j.value("host", host);
+    }
+    if (!invalid_keys.count("port")) {
+        port = j.value("port", port);
+    }
+    if (!invalid_keys.count("grpc_port")) {
+        grpc_port = j.value("grpc_port", grpc_port);
+    }
+    if (!invalid_keys.count("omp_threads")) {
+        omp_thread_count = j.value("omp_threads", omp_thread_count);
+    }
+    if (!invalid_keys.count("top_level_folder")) {
+        top_level_folder = j.value("top_level_folder", top_level_folder);
+    }
+    if (!invalid_keys.count("frontend_folder")) {
+        frontend_folder = j.value("frontend_folder", frontend_folder);
+    }
+    if (!invalid_keys.count("exit_timeout")) {
+        wait_time = j.value("exit_timeout", wait_time);
+    }
+    if (!invalid_keys.count("initial_timeout")) {
+        init_wait_time = j.value("initial_timeout", init_wait_time);
+    }
+    if (!invalid_keys.count("idle_timeout")) {
+        idle_session_wait_time = j.value("idle_timeout", idle_session_wait_time);
+    }
+}
+
+void ProgramSettings::ValidateJSON(const json& j) {
```

Comment:
Can we not turn the sets of keys into unordered maps, mapping the key values to pointers to the attributes, and set the attribute values here instead of doing it in a separate function?

~These collections could also be in this function's scope, since they are only used here. But maybe it's clearer to have them declared in the class scope.~ Sorry; I see they are also used in another place.

Reply:
Wouldn't that require the same amount of verbosity to define all map elements pointing to each different setting? Also, three different maps, one for each value type, same as with the keys. I agree that the code can be simplified, but it will look less clear.

Reply:
I think it will look much clearer to define the maps and iterate over the keys and values in a single set of loops both to validate the type and to set the value. We are halfway towards this solution anyway with the three sets of keys grouped by type. This will allow us to remove the additional pass over all the key values and attributes.

Reply:
To clarify, I am suggesting converting the sets of keys to maps of keys to attribute pointers and to use them once, not to create an additional set of maps and use them in a separate set of three loops.

Reply:
Thanks for explaining your points in detail. I implemented the changes as you suggest.